### PR TITLE
Add state machine, communication object, laps code

### LINKED ISSUE #76: Set up flight python project with state machine
## Description
Python3 function in its own module which will contain everything control-oriented for competition. Will receive communication object from the root project with a state attribute. Using this information it must be able to enter a state and perform the required actions. Will change state as competition goes on.

## Requirements
- [ ] Python3 function with communication object as parameter
- [ ] Isolated in its own module
- [ ] Research and implement a simple, yet effective, state machine model (no need for complexity)
- [ ] Must be able to enter a default state or resume a state upon first running the function
- [ ] States must be configurable and modular (easy to add new ones)

### LINKED ISSUE #85: Make communication object
## Description
This object will be owned by the root project and will be passed to vision and flight.

Flight and vision will fill it with continuous data.
For now, it will contain:
- The vision object located in `vision/interface.py` 
- A state field telling both vision and flight what our state is in the state machine.

## Requirements
- [x] Class in its own module
- [ ] Stores vision information
- [x] Stores state information
- [ ] Can mutate and access vision and state information simultaneously with no defect

### LINKED ISSUE #107: State files
From https://python-3-patterns-idioms-test.readthedocs.io/en/latest/StateMachine.html, implement a different file for each state in our current state diagram
image editing the diagram as needed
Place all files in `/flight/states`

### LINKED ISSUE #111: Pylon Laps State Implementation
Implement the state for doing laps in code so that when the state is triggered it will begin flying in the loop

Upon the state being interrupted and resumed it should be able to resume immediately, at the same exact location - so create more states as needed and utilize the communication object to facilitate that functionality

### PR DESCRIPTION
**Pull request by ClayMav**
Incorporates the work of John Shenouda, Elijah Wade, and Alex Becker to make our first full flight test, running competition code.
Closes #111, Closes #76, Closes #107, Closes #85

### CODE REVIEW

File: Pipfile [hunk 1]
```diff
@@ -0,0 +1,18 @@
+[[source]]
+name = "pypi"
+url = "https://pypi.org/simple"
+verify_ssl = true
+
+[dev-packages]
+pylint = "*"
+black = "*"
+
+[packages]
```
Comment (coledie) [inline Pipfile:10]: Should vision packages be in this? Is this a replacement for requirements.txt?

File: flight/states/Takeoff.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+            await drone.action.land()
+            return
+
+        return RunLaps()
+
+    async def wait_alt(self, drone: System):
+        previous_altitude = None
+        async for position in drone.telemetry.position():
+            altitude = round(position.relative_altitude_m, 2)
+            if altitude >= 2:
```
Comment (coledie) [inline flight/states/Takeoff.py:49]: Magic number?

File: flight/states/State.py [hunk 3]
```diff
@@ -0,0 +1,11 @@
+from mavsdk import System
+
+
+class State:
+    async def run(self, drone: System):
+        assert 0, "run not implemented"
```
Comment (coledie) [inline flight/states/State.py]:
```python
raise NotImplementedError("run not implemented")
```

File: flight/flight.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
+import asyncio
+import threading
+from mavsdk import System
+import mavsdk as sdk
+from .states import STATES, State
+
+
+class Comm:
+    state: str = "start"
```
Comment (coledie) [inline flight/flight.py]: Should this be replaced by the communication object in communication.py?

File: flight/flight.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+
+class StateMachine:
```
Comment (coledie) [inline flight/flight.py:8]: Is this a temporary replacement for states/State.py? or something new?

**Review by acbk2b — DISMISSED** [review]

File: flight/flight.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+SIM_ADDR: str = "udp://:14540"
+CONTROLLER_ADDR: str = "serial:///dev/ttyUSB0"
+
+
+async def print_flight_mode(drone: System) -> None:
+    """ Prints the flight mode when it changes """
+
+    previous_flight_mode: str = None
+
+    async for flight_mode in drone.telemetry.flight_mode():
```
Comment (coledie) [inline flight/flight.py:46]: This is super cool.

File: flight/states/RunLaps.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+                dx = math.copysign(
+                    35 * math.cos(math.asin(y / (math.sqrt((x ** 2) + (y ** 2))))), x
+                )
+                dy = math.copysign(
+                    35 * math.sin(math.asin(y / (math.sqrt((x ** 2) + (y ** 2))))), y
+                )
+
+            await drone.offboard.set_velocity_ned(sdk.VelocityNedYaw(dy, dx, alt, deg))
+
+            if abs(x) <= 10 and abs(y) <= 10:
```
Comment (coledie) [inline flight/states/RunLaps.py]: Magic numbers?

**Review by coledie — DISMISSED** [review]
Really just questions. Really cool stuff, nice work!

File: Pipfile (hunk 1 above)
Comment (ClayMav) [reply Pipfile:10]: They will need to be at some point, anything required for competition

File: flight/states/Takeoff.py (hunk 2 above)
Comment (ClayMav) [reply flight/states/Takeoff.py:49]: I'll have this fixed in a different PR focused on configuration

File: flight/flight.py (hunk 5 above)
Comment (ClayMav) [reply flight/flight.py:8]: Thats the state machine, then the states are separate

**Review by coledie — APPROVED** [review]

**Pull request merged**